src/compiler/AST/Common/Type.elm

```diff
@@ -30,32 +42,103 @@ getVarId type_ =
             Nothing
 
 
-toString : Type -> String
-toString type_ =
+empty : State
+empty =
+    State
+        { freeLetters = varsLetters
+        , mapping = Dict.empty
+        }
+
+
+toString : Type -> State -> ( String, State )
+toString type_ state =
     case type_ of
         Var int ->
-            "t" ++ String.fromInt int
+            getVariable int state
 
         Function t1 t2 ->
-            toString t1 ++ " -> " ++ toString t2
+            let
+                ( type1, state1 ) =
+                    wrapParens t1 state
+
+                ( type2, state2 ) =
+                    toString t2 state1
+            in
+            ( type1 ++ " -> " ++ type2
+            , state2
+            )
+
+        List param ->
+            wrapParens param state
+                |> Tuple.mapFirst ((++) "List ")
 
         Int ->
-            "Int"
+            ( "Int", state )
 
         Float ->
-            "Float"
+            ( "Float", state )
 
         Char ->
-            "Char"
+            ( "Char", state )
 
         String ->
-            "String"
+            ( "String", state )
 
         Bool ->
-            "Bool"
-
-        List param ->
-            "List " ++ toString param
+            ( "Bool", state )
 
         Unit ->
-            "()"
+            ( "Unit", state )
+
+
+dump : Type -> String
+dump type_ =
+    toString type_ empty
+        |> Tuple.first
```

Comment:
I'd either remove this or add a comment along the lines of "Don't use this for any serious work - use `toString` instead. For debugging only!"